libevmasm/Assembly.cpp

```diff
@@ -1422,6 +1436,13 @@ LinkerObject const& Assembly::assembleEOF() const
 
 	ret.bytecode += m_auxiliaryData;
 
+	auto appendedDataAndAuxDataSize = ret.bytecode.size() - dataStart;
+
+	// If some data was already added to data section we need to update data section refs accordigly
+	if (appendedDataAndAuxDataSize > 0)
+		for (auto [pos, val] : dataSectionRef)
+			setBigEndian(ret.bytecode, pos, 2, val + appendedDataAndAuxDataSize);
```

Comment:
It looks that during all the splitting and rebasing one important code section was missed. The fragment calculates the maximum offset accessed by the `DataLoadN` and adjust the `predeploy_data_size` (data section size value in the header) accordingly. We need definitely better tests also.

Reply:
Yes I agree with different name. Any idea for the name? :) Moreover it's also related to my comment here https://github.com/ethereum/solidity/pull/15456#discussion_r1777278126

Reply:
> We need definitely better tests also.

True, but for now I'd not worry about it *too* much. We need *something* and we definitely need coverage for things where there's a risk of non-EOF being affected, but the idea with EOF is too basically speed through it to the point where we can run semantic tests. Once we get them to pass, we'll have some assurance that things can't be too wrong and we'll be able to focus on testing and ironing out specific parts.

So generally, I'll definitely be pointing out spots where I think something is wrong, but also it is not that important to get them all right at this point. We're fine with it being unstable and incomplete to some extent.

Reply:
> Any idea for the name?

Maybe we should take a step back and talk about the design of this instruction first. Actually, what kinds of data we can have now? This is my current understanding:
1. Explicit Yul data - comes from `data` sections other than `".metadata"` in a Yul object.
1. Implicit Yul data - added in transition from Yul to assembly. E.g. long strings that in Yul are just literals, but become data when assembled.
1. Metadata - comes from the `".metadata"` data section in a Yul object.
1. Static auxdata - comes from `RETURNCONTRACT`, not known at compilation time.
1. Dynamic auxdata - comes from `RETURNCONTRACT`, not known at compilation time and inaccessible via `DATALOADN`.

AFAIK metadata ends up in `m_auxiliaryData` and implicit+explicit Yul data are both in `m_data`. Static auxdata at compilation time is not in the binary at all.

It seems that your idea is to allow the user to access only (4). In that case `auxdataloadn()` would be a descriptive name.

We may still want to add `dataloadn()` in some form in addition to it, to make it possible to access named Yul data sections that way. Probably not `dataloadn(offset)` since it needs a literal (so expressions like `dataload(add(dataoffset('mydatasection'), 42))` would not work), but maybe `dataloadn(data_name, offset)`? The extra argument should make it clear that the offset is not the same as in the EVM opcode.

Also, `auxdataoffset()` and `staticauxdatasize()` might be useful in combination with `datacopy()`, but we could leave those for later. I was originally going to suggest that you could do things like `dataloadn(add(auxdataoffset(), 42))` to uniformly access auxdata and named data sections, but then realized that we would not be able to use expressions with it.

Reply:
> Finally, this is missing a validation that when you add `appendedDataAndAuxDataSize`, the value still does not exceed `0xffff`. There is an assert in `setBigEndianUint16()`, but this looks like something that can actually happen in normal usage and needs a runtime check. The user has no direct control over the size of data appended by the compiler.

`setBigEndian(ret.bytecode, refPosition, 2, ...)` checks that the value  does not exceed `0xffff`.  The template parameter `ValueT` is size_t in this case. It should be fine imo.

Reply:
`setBigEndian()` *assumes* that it does not exceed `0xffff`. It's not a validation. If it's not true, we'll get a compiler crash, spewing diagnostic info. If it can actually happen, then we need a proper validation with something like `solRequire()` that will actually be handled as a proper codegen error.

Reply:
Ok. I get it now. Fixed.